nomad/state/state_store.go

```diff
@@ -1814,6 +1859,260 @@ func (s *StateStore) VaultAccessorsByNode(ws memdb.WatchSet, nodeID string) ([]*
 	return out, nil
 }
 
+// UpsertDeploymentStatusUpdate is used to upsert deployment status updates and
+// potentially make a evaluation
+func (s *StateStore) UpsertDeploymentStatusUpdate(index uint64, req *structs.DeploymentStatusUpdateRequest) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	if err := s.upsertDeploymentStatusUpdateImpl(index, req.DeploymentUpdate, txn); err != nil {
+		return err
+	}
+
+	// Upsert the job if necessary
+	if req.Job != nil {
+		if err := s.upsertJobImpl(index, req.Job, txn); err != nil {
+			return err
+		}
+	}
+
+	// Upsert the optional eval
+	if req.Eval != nil {
+		if err := s.nestedUpsertEval(txn, index, req.Eval); err != nil {
+			return err
+		}
+	}
+
+	txn.Commit()
+	return nil
+}
+
+// upsertDeploymentStatusUpdateImpl is used to upsert deployment status updates
+func (s *StateStore) upsertDeploymentStatusUpdateImpl(index uint64, u *structs.DeploymentStatusUpdate, txn *memdb.Txn) error {
+	// Retrieve deployment
+	ws := memdb.NewWatchSet()
+	deployment, err := s.deploymentByIDImpl(ws, u.DeploymentID, txn)
+	if err != nil {
+		return err
+	} else if deployment == nil {
+		return fmt.Errorf("Deployment ID %q couldn't be updated as it does not exist", u.DeploymentID)
```

Comment:
It's a bit confusing because it's upserting a field/subobject on deployments, so the deployment must exist but doesn't need a prior status set?

+1 to calling it an Update :)